lighthouse-core/test/gather/gatherers/trace-elements-test.js

```diff
@@ -268,4 +292,18 @@ describe('Trace Elements gatherer - GetTopLayoutShiftElements', () => {
     const total = sumScores(result);
     expectEqualFloat(total, 2.5);
   });
+
+  it('gets animated node ids without duplicates', () => {
```

Comment:
it'd be nice to add a test that exercises the bulk of the logic we're mutating in the gatherer here with the combination of multiple nodeId types

looks like `afterPass` tests were missing in the first go around, bummer you'll have to start from scratch :/

trace-processor-test has a few good examples of using `createTestTrace` while pushing some custom events

https://github.com/GoogleChrome/lighthouse/blob/439596e9cc5e767864b2ab29ba21782246a3df1d/lighthouse-core/test/lib/tracehouse/trace-processor-test.js#L355-L363

Reply:
Creating an `afterPass` test will require some kind of phony `driver` to simulate `driver.resolveNodeIdToObjectId()` and `driver.sendCommand()`.

Would the solution here be to simulate these functions with hardcoded results for each node id / object id?

Reply:
Correct, you'll need to stub out those. We haven't done a very good job in the past of doing this in gatherer tests since most of them were written before we had a framework for mocking the driver well, but since then we have a pretty nice mockCommand approach in the driver tests you can follow

https://github.com/GoogleChrome/lighthouse/blob/92b7c7642ad90f9cfd9231973dee7ef04675138e/lighthouse-core/test/gather/driver-test.js#L137

you would hardcode the node id and object id for these. I'm not looking for anything enormous or complex, just *something* that checks it tries to find each of the 3 node types and merges them with the correct labels, so something roughly like

```js
  it('identifies the node ID', async () => {
    connectionStub.sendCommand = createMockSendCommandFn()
      .mockResponse('DOM.resolveNode', {object: {objectId: 123}})
      .mockResponse('Runtime.callFunctionOn', {result: {value: {all: 'the', properties: 1}}})
      .mockResponse('DOM.resolveNode', {object: {objectId: 124}})
      .mockResponse('Runtime.callFunctionOn', {result: {value: {all: 'the', properties: 1}}})
      .mockResponse('DOM.resolveNode', {object: {objectId: 125}})
      .mockResponse('Runtime.callFunctionOn', {result: {value: {all: 'the', properties: 1}}});

    const fakeTrace = createFakeTestTrace();
    const driver = new Driver(connectionStub);
    const result = await TraceElementsGatherer.afterPass({driver}, {trace: fakeTrace});
  });
```